server/src/auth.rs

```diff
@@ -0,0 +1,144 @@
+use base64::decode;
+use oxhttp::model::{HeaderName, Request, Status};
+
+pub enum Store {
+    All,
+    // this would allow specifying different permissions for each store,
+    // not 100% sure if i understood this correctly,
+    // or if it's easier to leave this out.
+    Defined(String),
+}
+
+// TODO: rename to Action?!
+pub enum Permission {
+    StaticFiles,
+    Query,
+    Update(Store),
+    Create(Store),
+    Delete(Store),
+}
+
+#[derive(Eq, PartialEq, Debug)]
+struct NamedUser {
+    username: String,
+    password: String,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+enum User {
+    Anonymous,
+    Named(NamedUser),
+}
+
+// TODO: next line is c&p from main.rs
+type HttpError = (Status, String);
+
+fn is_valid_user(user: &NamedUser) -> bool {
+    // this is a dummy implementation, needs some kind of config backend:
+    // i thought of an additional (read only?) store you can specify on the
+    // command-line: the store should contain, user, hashed passwords, and a
+    // link between users ans permissions/actions. but not 100% sure how to
+    // model this in rdf.
+    return user.username == "admin" && user.password == "password";
+}
+
+fn can_user_access(user: &User, required_permissions: Permission) -> bool {
+    // see is_valid_user
+    match required_permissions {
+        Permission::Query => match user {
+            User::Anonymous => false,
+            User::Named(_) => true,
+        },
+        _ => true,
+    }
+}
+
+// a similar function could be moved to oxhttp?
+fn read_user_from_request(request: &Request) -> anyhow::Result<User> {
+    match request.header(&HeaderName::AUTHORIZATION) {
+        None => Ok(User::Anonymous),
+        Some(header) => match header.to_str().unwrap().split_once(" ") {
+            None => Err(anyhow::anyhow!("Could not read Header")),
+            Some((prefix, payload)) => {
+                if prefix != "Basic" {
+                    return Err(anyhow::anyhow!("Only support Basic Auth"));
+                }
+                match decode(payload) {
+                    Err(_) => Err(anyhow::anyhow!("Could not base64 decode header")),
```

Comment:
anyhow has the great [`context`](https://docs.rs/anyhow/latest/anyhow/trait.Context.html#tymethod.context) method that allows to add context to errors.

For example you could write
```rust
let decoded = decode(payload).context("Could not base64 decode header")?
```
and get the same behavior + the original error is tracked and printed after the new message.